### PR TITLE
AdEMAMix Optimizer

### PR DESCRIPTION
**Pull request by mpagli**
Example of arguments:

```python
REGULARIZATION_ARGS=(
	--attention-dropout 0.0
	--hidden-dropout 0.0
	--weight-decay 0.1
	--clip-grad 0.1
	--adam-beta1 0.9
	--adam-beta2 0.95
	--ademamix-alpha 8
	--ademamix-beta3 0.9999
	--ademamix-beta3-warmup 100000
	--ademamix-alpha-warmup 100000
)

TRAINING_ARGS=(
	--micro-batch-size 1
	--global-batch-size $GBS
	--train-iters 100000
	--log-interval 1
	--cross-entropy-loss-fusion
	--disable-bias-linear
	--optimizer ademamix
	--dataloader-type single
	--sequence-parallel
	--manual-gc
	--manual-gc-interval 5000
)
```

Results for 8B models so far:

image

### CODE REVIEW

**Comment by ischlag** [comment]
running baseline llama3 1b with WSD and cosine decay schedule. See: [](https://wandb.ai/ischlag/optimiser-ablation?nw=nwuserischlag)

**Comment by haeggee** [comment]
something i wanted to say yesterday already: why would you actually want to get rid of warmup? sure, it might work, but making training without warmup work well is ongoing research (https://arxiv.org/pdf/2410.23922), and degradation might only show up late in a run. especially for big/long runs, why risk it if it doesn't outperform? we can keep a few thousand warmup steps

**Comment by ischlag** [comment]
image

Because it is simpler and it doesn't help up to 1b (maybe even larger). Looking at these ablation runs here (wandb link below) you can see that warmup actually hurts stability (weavy lines are with warmup; other without). So if it seems more stable and performs better than why not remove it? 

That said, it doesn't matter so much. Just like init range, after enough steps the runs look the same. 

https://wandb.ai/ischlag/wsd-ablation?nw=nwuserischlag&panelDisplayName=lm+loss&panelSectionName=Charts

**Comment by mpagli** [comment]
In all the experiments we did with AdEMAMix and WSD, we used warmup (2k or 3k). We did not have to change any other parameters when switching between cosine and WSD. Concerning the decay duration, the longer the better, but is should work with using the same number of iterations as Adam. 

In the paper we have the following when using a decay duration of 100k steps with 1.3B models:
image

If we compare with doing 200k steps:
image

We see that Adam is better with the longer cooldown, and it reduces the gap between the two AdEMAMix runs. 

100k or 200k might seem like a lot, keep in mind that the batch size is relatively small in those experiments (128*1024). Recently, @Andron00e did more experiments in which we decay for fewer iterations without problems:

image
In the above plot, we use a 124M model, the token labels on the x-axis correspond to 8k,16k,32k,48k,64k,128k

So, overall, when using WSD with AdEMAMix, there is not much to change. Keep everything as for cosine, if the decay period is longer it might be better, but it should also work with shorter decay.

**Comment by ischlag** [comment]
I pushed the most recent commits and configs @mpagli to this branch. @mpagli can you please create copies of that you think we should try. First evaluate them at 300m. 

We run at total of 48441 steps to cover 100B tokens. Our batch size is 2M tokens so quite a bit more from your ~130k.

**Comment by ischlag** [comment]
I moved the discussion about hyperparameters and the submission scripts to our [pretrain repo](https://github.com/swiss-ai/pretrain/issues/70). We should move the hyperparam discussion over there. 

I think we should keep the discussion here strictly about what is necessary to merge ademamix into megatron. For now, it works for us to run experiments on this branch but to merge it into this megatron-lm fork we have to make sure the optimiser works properly with all components of megatron.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 1]
```diff
@@ -487,9 +489,20 @@ def __init__(
         for model_chunk in self.model_chunks:
             assert self.ddp_config == model_chunk.ddp_config
 
-        assert (
-            isinstance(optimizer, Adam) or optimizer is None
-        ), "Only Adam currently supported, due to checkpointing requirements."
+        ## assert isinstance(
+        ##    optimizer, Adam
+        ## ), "Only Adam currently supported, due to checkpointing requirements."
```
Comment (TJ-Solergibert) [inline megatron/core/optimizer/distrib_optimizer.py:494]:
```suggestion
```

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -487,9 +489,20 @@ def __init__(
...
+        if isinstance(optimizer, Adam):
+            self.optimizer_name = 'adam'
+            self.optimizer_keys = ("param", "exp_avg", "exp_avg_sq")
+        elif isinstance(optimizer, AdEMAMix):
+            self.optimizer_name = 'ademamix'
+            if config.adam_beta1 != 0.0:
+                self.optimizer_keys = ("param", "exp_avg_slow", "exp_avg_fast", "exp_avg_sq")
+            else:
+                self.optimizer_keys = ("param", "exp_avg_slow", "exp_avg_sq")
```
Comment (TJ-Solergibert) [inline megatron/core/optimizer/distrib_optimizer.py:503]:
```suggestion
            self.optimizer_keys = ("param", "exp_avg_slow", "exp_avg_sq")
            if config.adam_beta1 != 0.0:
                self.optimizer_keys = ("param", "exp_avg_slow", "exp_avg_fast", "exp_avg_sq")
```

File: megatron/core/optimizer/distrib_optimizer.py [hunk 3]
```diff
@@ -693,8 +706,14 @@ def load_state_dict(self, state_dict):
...
-
-                            tensors = {"exp_avg": init_shard(), "exp_avg_sq": init_shard()}
+                            
+                            if self.optimizer_name == 'adam':
+                                tensors = {"exp_avg": init_shard(), "exp_avg_sq": init_shard()}
+                            elif self.optimizer_name == 'ademamix':
+                                if len(self.optimizer_keys) == 4: # beta1 != 0
+                                    tensors = {"exp_avg_slow": init_shard(), "exp_avg_fast": init_shard(), "exp_avg_sq": init_shard()}
+                                else: # beta1 == 0
+                                    tensors = {"exp_avg_slow": init_shard(), "exp_avg_sq": init_shard()}
```
Comment (TJ-Solergibert) [inline megatron/core/optimizer/distrib_optimizer.py:716]:
```suggestion
                                tensors = {"exp_avg_slow": init_shard(), "exp_avg_sq": init_shard()}
                                if "exp_avg_fast" in self.optimizer_keys:
                                    tensors["exp_avg_fast"] =  init_shard()
```

**Review by TJ-Solergibert — COMMENTED** [review]
Hi! We should check how we handle `HAVE_APEX_OR_TE` in `megatron/core/optimizer/distrib_optimizer.py`, e.g. https://github.com/swiss-ai/Megatron-LM/blob/main/megatron/core/optimizer/distrib_optimizer.py#L609.

Because of our environment (NGC PyTorch images) it will always be `True`, but precisely this global is used to differentiate between the Apex/TransformerEngine optimizer and the regular PyTorch optimisers (This case). Maybe just setting this global to False after checking the optimizer name is sufficient.

**Comment by mpagli** [comment]
Thanks for the cleanup. 

I did a quick test to see if setting beta1=0 properly saves memory. It works:
image

**Review by TJ-Solergibert — APPROVED** [review]
Added little comment clarifying that AdEMAMix stores the step the same way as APEX & TE Fused Adam optimisers. Thanks @mpagli
